cpp/ql/src/semmle/code/cpp/security/implementation/SafeExternalAPIFunction.qll

```diff
@@ -0,0 +1,15 @@
+/**
+ * Provides a class for modeling external functions that are "safe" from a security perspective.
+ */
+
+private import cpp
+
+/**
+ * A `Function` that is considered a "safe" external API from a security perspective.
+ */
+abstract class SafeExternalAPIFunction extends Function { }
+
+/** The default set of "safe" external APIs. */
+private class DefaultSafeExternalAPIFunction extends SafeExternalAPIFunction {
+  DefaultSafeExternalAPIFunction() { this.hasGlobalName(["strcmp", "strlen", "memcmp"]) }
```

Comment:
There's already this fairly good list of functions known to be "pure": https://github.com/github/codeql/blob/7f0ad2d2320bcec781cd60e02934471238817d73/cpp/ql/src/semmle/code/cpp/models/implementations/Pure.qll#L9-L11

Rather than listing them here again, perhaps we can reuse that list of functions. That could be either directly or using a new model interface to describe functions that have no side effects other than their modelled memory side effects. Oh, and shouldn't we respect `parameterIsAlwaysReturned` in the data flow libraries?

Reply:
> Rather than listing them here again, perhaps we can reuse that list of functions. That could be either directly or using a new model interface to describe functions that have no side effects other than their modelled memory side effects.

Good point. It should be noted that all those pure functions have dataflow models, so they're already being excluded from `ExternalAPIDataNode`.
So the new model interface you're suggesting would be a `PureFunction` model that doesn't necessarily carry taintflow, but is otherwise very much like `PureStrFunction` (except that it can include non-string functions). Is that correct?

Reply:
> Oh, and shouldn't we respect `parameterIsAlwaysReturned` in the data flow libraries?

We could do that, yes! I guess we can add a case for `AliasFunction` (and not just `DataFlowFunction`) in `DataFlowUtil`'s `modelFlow` predicate, but a quick review of all the models where `parameterIsAlwaysReturned(n)` holds for some `n` it seems like they also satisfy `hasDataFlow(input, output)` where `input.isParameter(n)` and `output.isReturnValue()`, so I don't think this would make any difference.

Reply:
I've added a new model class in 715f2333609c65fae9920fa038ed9650c4167ecd. I also added very brief QLDocs to the classes in `Pure.qll` to avoid PR-check complaints (although now I see that it's in an implementation directory, so maybe that wasn't necessary).

Reply:
That might have a semantic conflict with #4634  and definitely has a philosophical conflict with it. I don't think we have an agreement on how to manage classes that have model implementations but would also be useful as sets in other contexts.

Reply:
Good point. I'm not sure what the best approach is. I'm not sure we can come up with a good model interface that `DefaultSafeExternalAPIFunction` can use in place of the implementation classes (like @geoffw0 did for four outliers in #4634).
One possible interface is one that describes pure functions, but we already have an implementation model with that name (and then we have to start a philosophical discussion on what it means to be pure).

Reply:
I agree there's likely to be a conflict. In terms of timing, this PR gets to go first because it's a more urgent priority. We'll have to discuss further and fix any fallout in #4634.